src/Avalonia.Controls/Generators/TreeItemContainerGenerator.cs

```diff
@@ -101,24 +102,50 @@ protected override IControl CreateContainer(object item)
         public override IEnumerable<ItemContainerInfo> Clear()
         {
             var items = base.Clear();
-            Index.Remove(0, items);
+            Index?.Remove(0, items);
             return items;
         }
 
         public override IEnumerable<ItemContainerInfo> Dematerialize(int startingIndex, int count)
         {
-            Index.Remove(startingIndex, GetContainerRange(startingIndex, count));
+            Index?.Remove(startingIndex, GetContainerRange(startingIndex, count));
             return base.Dematerialize(startingIndex, count);
         }
 
         public override IEnumerable<ItemContainerInfo> RemoveRange(int startingIndex, int count)
         {
-            Index.Remove(startingIndex, GetContainerRange(startingIndex, count));
+            Index?.Remove(startingIndex, GetContainerRange(startingIndex, count));
             return base.RemoveRange(startingIndex, count);
         }
 
         public override bool TryRecycle(int oldIndex, int newIndex, object item) => false;
 
+        public void UpdateIndex()
+        {
+            if (Owner is TreeView treeViewOwner && Index == null)
+            {
+                Index = new TreeContainerIndex();
+                _treeView = treeViewOwner;
+            }
+            else if (Owner.IsAttachedToLogicalTree)
+            {
+                var treeView = Owner.GetSelfAndLogicalAncestors().OfType<TreeView>().FirstOrDefault();
```

Comment:
@grokys I wonder if we could de-LINQ this?

Reply:
We could but it's usually run only once, on tree attachment, and all this code will go away soon when we use `ItemsRepeater` for items presenters...

Reply:
As a heads up: I am working on a replacement for most of the visual tree ancestor/descendant traversals. Simply because iterator blocks and LINQ are around 10x slower than regular traversal.

Reply:
So in the new API this might look like `VisualTreeOperations.FindAncestorOfType<TreeView>(Owner)`.